finished all todos - no bonus

ZiqianPan opened the pull request.

ZiqianPan committed:
finished all todos

mikeboyle commented on src/App.js:
```diff
@@ -1,15 +1,67 @@
+  if (searchInput) {
+    dataToDisplay = userData.filter((user) => {
+      const { name, company, country } = user;
+      let lowerCaseSearch = searchInput.toLowerCase();
```
👍 Nice! Saves a bit of processing time by only lowercasing once.

mikeboyle commented on src/components/Users/Users.js:
```diff
@@ -1,12 +1,30 @@
+
+const Users = ({ users, setExpanded, expanded }) => {  
+  const handleExpanded = (id) => {
+    let copyExpaned = expanded;
```
I think it's unnecessary to use `copyExpanded` here, for a couple of reasons.

1. `copyExpanded = expanded` doesn't actually copy it. Both `copyExpanded` and `expanded` point to the same array; no copy is made.
2. In both the `if` and `else` blocks, you use safe non-mutating methods that return a new array (`filter` and `[...copyExpanded, id]`)

Even if `copyExpanded` was a true copy, this would be overkill because in 2) we see that you are already safely handling the original array.

mikeboyle commented on src/components/User/User.js:
```diff
@@ -15,13 +15,17 @@ const User = ({ user }) => {
-          <p>{about}</p>
-        </div>
+
+        {expanded.includes(id) && (
```
See related comment in `<Users />`

mikeboyle commented on src/components/Users/Users.js:
```diff
@@ -1,12 +1,30 @@
+            user={user}
+            handleExpanded={handleExpanded}
+            expanded={expanded}
+            setExpanded={setExpanded}
```
I think you can delete this line. The `<User />` component does not need the `setExpanded` update function and also doesn't do anything with this prop.

mikeboyle commented on src/components/Users/Users.js:
```diff
@@ -1,12 +1,30 @@
+            key={id}
+            user={user}
+            handleExpanded={handleExpanded}
+            expanded={expanded}
```
Passing the entire `expanded` array to a single `<User />` works in this implementation because you have the User check if its id is in the array. However, this is not the approach I would take personally. I would rather do that check for `expanded.includes(id)` here in `<Users />` and pass a simple boolean prop to `<User />`.

The reason why I prefer this approach is because the main idea of lifting state up is to give total control and responsibility for the state back to the parent. This means it's the parent's responsibility to figure out if the child should be expanded. In the code here, the parent is passing the entire array down to the child and making the child responsible for figuring it out. I think it would be better for the parent to take full control for this.

In addition, it's just a good idea in software design in general to make sure that each part of the system has a single, clearly defined responsibility. Another way of looking at that is: for all of the things that the system has to do, we want there to be only one part of the system that is fully responsible for it. When you design your app like this, it will be much easier to debug, refactor and extend its functionality over time.

mikeboyle reviewed (commented):
Looking great! I have a couple of small concerns about how the parent and child divide up responsibility. But this works and the state has been lifted up to the right places. Nice job!

ZiqianPan committed:
took suggestions

ZiqianPan committed:
Lifed up expaned.includes(id) into parent

I have moved the expanded.includes(id) from the children
up into the parent when it is passing down the props. This
way the parent is doing all the work / responsible for the children
